### PR TITLE
Testbench and some reorganization

### PR DESCRIPTION
**Pull request by telegraphic**
Hey all,

Here's a few testbenches that I think would work well as tutorials too. There's a few things going on here (sorry):

* I created a directory called 'blocks', in which I propose we move all blocks. At the moment I've added relative blocks via imports in the blocks/__init__.py script

* I've created a sink and source block for reading/writing binary data using numpy's fromfile() and tofile() commands. 

* I made a very basic FFT block that wraps the bfCUDA FFT. 

* I added a testbench directory, which can be used for running basic end-to-end pipeline tests and is also good as self-documentation.

* I wrote a testbench for the binary file read + write blocks. This works well.

* I wrote a testbench for the FFT, using which I get odd results with the testbench that indicate something is broken. I wonder if there's a device synchronize problem in the fft.cu code?

* I also added four very basic cuda kernels for basic operations (+ -  / * ). There is no doubt a cooler way to do this. Adding this is all part of my goal of making bifrost a builder's toolkit a la CASPER.

### CODE REVIEW

File: python/bifrost/fft_block.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from pipeline import TransformBlock
+from copy import deepcopy
+from fft import fft
+
+class FFTBlock(TransformBlock):
```
Comment (MilesCranmer) [inline python/bifrost/fft_block.py]: I wonder if we should have this block handle both CPU and GPU FFT's, or maybe rename one of them so there is no conflict.
Thoughts?

**Comment by MilesCranmer** [comment]
Great work Danny, this looks good to me! I will test the code out now.

File: python/bifrost/fft_block.py (hunk 1 above)
Comment (benbarsdell) [inline python/bifrost/fft_block.py]: Ideally it should look at what space the inputs and outputs are to decide between CPU and GPU.

I've been working on a new FFT block that will add some features, but it's proving a bit more complicated than I thought.

Comment (telegraphic) [inline python/bifrost/fft_block.py]: Agreed, that would be nice. Can CPU/GPU select be done once at instantiation?

No strong opinion on if we should merge examples and testbench. Maybe testbench and tutorials would be a better split?

**Comment by MilesCranmer** [comment]
This looks good for merging!

Maybe after, we could follow your idea and reorganize the block directories? We'll also need to rewrite some unit tests (and other files, e.g., example pipelines) to import correctly.

**Comment by MilesCranmer** [comment]
(I just realized that Ben created a commit which did this already)

File: docs/source/your-first-blocks.rst [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+            
+        def on_data(self, ispan, ospan):
+            in_nframe  = ispan.nframe
+            out_nframe = in_nframe
+            
+            idata = ispan.data 
+            odata = ospan.data
+            
+            odata[...] = idata + self.n_to_add
+            return out_nframe
```
Comment (benbarsdell) [inline docs/source/your-first-blocks.rst]: You don't actually need the return statement if you processed the whole span :)

**Comment by telegraphic** [comment]
Now updated with more documentation

**Comment by MilesCranmer** [comment]
Looks good to me. Merging.

**Comment by MilesCranmer** [comment]
By the way, you could also tell travis to execute the new unit tests you added if they are all non-GPU.

**Pull request merged**